### PR TITLE
[15.0] edi_storage: path configurable by type and param

### PR DESCRIPTION
**Pull request by jcoux**
Backported on https://github.com/OCA/edi/pull/578

By setting specific storage path on exchange type advanced settings, we
can configure specific path for each exchange type (input or output).

Path can also be configured by config parameters to allow to easily set
different values depending of the environment.

### CODE REVIEW

File: edi_storage_oca/models/edi_backend.py [hunk 1]
```diff
@@ -145,16 +145,24 @@ def _storage_create_record_if_missing(self, exchange_type, remote_file_name):
...
         if not exchange_type.exchange_filename_pattern:
-            # If there is not patter, return everything
-            return self.storage_id.list_files(self.input_dir_pending)
+            # If there is not pattern, return everything
+            files = self.storage_id.list_files(full_input_dir_pending)
+            filenames = []
+            for file in files:
+                if file.strip("/"):
+                    filenames.append(file)
+            return filenames
```
Comment (simahawk) [inline edi_storage_oca/models/edi_backend.py]:
```suggestion
            filenames = [x for x in self.storage_id.list_files(full_input_dir_pending) if x.strip("/")]
            return filenames
```
Wondering: when do you get `/` only?

File: edi_storage_oca/models/edi_exchange_type.py [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
+
+from odoo import api, fields, models
+
+from odoo.addons.base_sparse_field.models.fields import Serialized
 
 
 class EDIExchangeType(models.Model):
     _inherit = "edi.exchange.type"
 
+    storage_settings = Serialized(default={}, compute="_compute_storage_settings")
```
Comment (simahawk) [inline edi_storage_oca/models/edi_exchange_type.py]: ```help="Alias to `storage` key in type settings"```

File: edi_storage_oca/models/edi_exchange_type.py [hunk 3]
```diff
@@ -19,3 +25,44 @@ class EDIExchangeType(models.Model):
...
+            icp = self.env["ir.config_parameter"].sudo()
+            path = icp.get_param(path_config_param)
+            if path:
+                return PurePath(path)
+
+    def _get_full_exchange_type_path(self, directory=None):
+        self.ensure_one()
+        path_prefix = self._get_exchange_type_path()
+        path = PurePath((directory or "").strip().rstrip("/"))
+        if path_prefix:
```
Comment (simahawk) [inline edi_storage_oca/models/edi_exchange_type.py:58]: isn't this the same done in `_make_remote_file_path`?

File: edi_storage_oca/models/edi_exchange_type.py [hunk 4]
```diff
@@ -19,3 +25,44 @@ class EDIExchangeType(models.Model):
...
+        if path:
+            return PurePath(path)
+        path_config_param = self.storage_settings.get("path_config_param")
+        if path_config_param:
+            icp = self.env["ir.config_parameter"].sudo()
+            path = icp.get_param(path_config_param)
+            if path:
+                return PurePath(path)
+
+    def _get_full_exchange_type_path(self, directory=None):
```
Comment (simahawk) [inline edi_storage_oca/models/edi_exchange_type.py]:
```suggestion
    def _storage_fullpath(self, directory=None):
```

File: edi_storage_oca/models/edi_exchange_type.py [hunk 5]
```diff
@@ -19,3 +25,44 @@ class EDIExchangeType(models.Model):
         "the files to be fetched from the pending directory in the related "
         "storage. E.g: `.*my-type-[0-9]*.\\.csv`"
     )
+
+    @api.depends("advanced_settings")
+    def _compute_storage_settings(self):
+        for rec in self:
+            rec.storage_settings = rec.advanced_settings.get("storage", {})
+
+    def _get_exchange_type_path(self):
```
Comment (simahawk) [inline edi_storage_oca/models/edi_exchange_type.py]:
```suggestion
    def _storage_path(self):
```

**Review by simahawk — CHANGES_REQUESTED** [review]

File: edi_storage_oca/models/edi_backend.py (hunk 1 above)
Comment (jcoux) [reply edi_storage_oca/models/edi_backend.py]: On this unit test:
https://github.com/OCA/edi/blob/15.0/edi_storage_oca/tests/test_edi_backend_storage.py#L206
The `list_files` function returns:
`['/', '/some-file.csv', '/another-file.csv']`

File: edi_storage_oca/models/edi_exchange_type.py (hunk 3 above)
Comment (jcoux) [reply edi_storage_oca/models/edi_exchange_type.py:58]: Just removed `_make_remote_file_path` to avoid duplicates parts of code.

File: edi_storage_oca/components/base.py [hunk 6]
```diff
@@ -40,15 +43,15 @@ def _dir_by_state(self, direction, state):
...
-        :param state: string stating state of the exchange
-        :param filename: string for file name
-        :return: PurePath object
-        """
-        return self._dir_by_state(direction, state) / filename.strip("/ ")
+    def _get_remote_file_path(self, state, filename=None):
+        """Retrieve remote path for current exchange record."""
+        filename = filename or self.exchange_record.exchange_filename
+        direction = self.exchange_record.direction
+        directory = self.backend[direction + "_dir_" + state] or ""
```
Comment (simahawk) [inline edi_storage_oca/components/base.py]: pls keep using `self._dir_by_state` here which is what you are doing here but w/out validation

File: edi_storage_oca/tests/test_components_base.py [hunk 7]
```diff
@@ -18,14 +18,18 @@ def test_remote_file_path(self):
...
 
-        with self.assertRaises(AssertionError):
-            self.checker._remote_file_path("WHATEVER", "error", "foo.csv")
+        with self.assertRaises(ValueError):
+            self.checker.exchange_record.direction = "WHATEVER"
+            self.checker._get_remote_file_path("error", "foo.csv")
 
-        with self.assertRaises(AssertionError):
-            self.checker._remote_file_path("input", "WHATEVER", "foo.csv")
+        with self.assertRaises(KeyError):
```
Comment (simahawk) [inline edi_storage_oca/tests/test_components_base.py]: key error?

File: edi_storage_oca/components/base.py [hunk 8]
```diff
@@ -40,15 +43,15 @@ def _dir_by_state(self, direction, state):
...
+        path = self.exchange_record.type_id._storage_fullpath(
+            directory=directory, filename=filename
+        )
+        return path
```
Comment (simahawk) [inline edi_storage_oca/components/base.py:68]: why not having always PurePath objects till the end as it was before?

File: edi_storage_oca/tests/common.py [hunk 9]
```diff
@@ -76,13 +76,13 @@ def _file_fullpath(
         record = record or self.record
         if not fname:
             fname = self._filename(record, ack=ack)
-        if not direction:
-            direction = record.direction
+        if direction:
```
Comment (simahawk) [inline edi_storage_oca/tests/common.py]: why this change? Is not the same thing as before. Also, the direction on the record cannot be changed because it's a related from the type.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: edi_storage_oca/components/base.py (hunk 6 above)
Comment (simahawk) [reply edi_storage_oca/components/base.py]: we should probably keep the old method and log deprecation in case someone is still using it

**Comment by jcoux** [comment]
@simahawk All is done: you can update your review ;)

**Review by damdam-s — APPROVED** [review]

**Review by SilvioC2C — APPROVED** [review]
LGTM overall

**Review by simahawk — APPROVED** [review]
